Added docs for OVH managed kubernetes (#74)
elastisys/compliantkubernetes: docs/operator-manual/ovh-managed-kubernetes.md

# Compliant Kubernetes Deployment on OVH Managed Kubernetes

This document contains instructions on how to setup a service cluster and a workload cluster in [OVH](https://www.ovh.com).
The following are the main tasks addressed in this document:

1. Setting up Compliant Kubernetes for OVH Managed Kubernetes
1. Deploying Compliant Kubernetes on top of two kubernetes clusters.

The instructions below are just samples, you need to update them according to your requirements.

Before starting, make sure you have all necessary tools.

## Setup

Create two kubernetes clusters in OVH, follow [this guide](https://docs.ovh.com/gb/en/kubernetes/creating-a-cluster/).

> For the service cluster you can start with creating 3 nodes of size `B2-7` and add a `B2-15` node.

Reviewer: Awesome! We should add "sizing hints" in all supported platforms. Could you make it stand a bit out, so we remember? Maybe use an [admonition](https://squidfunk.github.io/mkdocs-material/reference/admonitions/) like `!!!note "Sizing Hint`.

The workload cluster is fine with 3 `B2-7` nodes.

### Configure Compliant Kubernetes

Start by preparing your shell with some variables that will be used.

```ShellSession

> REGION="waw" # Region for the cluster

Reviewer: Can you give a usable example in Europe in the comment?

Author: This is a usable example in central europe (Warsawa)

Reviewer: :joy: I thought `waw` was a placeholder like `blah` or `foo` or `bar`. :joy:

> ISSUER_MAIL="user@example.com" # Mail that will be used for the issuer

Reviewer 2: ```suggestion
ISSUER_MAIL="user@example.com" # Mail that will be used for the LetsEncrypt certificate issuer
```

export CK8S_CONFIG_PATH=~/.ck8s/my-ovh-cluster # Path for the configuration
export CK8S_ENVIRONMENT_NAME=my-ovh-cluster # Name of the environment

> export CK8S_PGP_FP="FOOBAR1234567" # Fingerprint of your PGP key

Reviewer: ```suggestion
export CK8S_PGP_FP=<your GPG key fingerprint>  # retrieve with gpg --list-secret-keys
```

Author: I'm starting to dislike the `<Description of stuff>` but rather use something like this and a comment instead.
Since the example isn't really copy-paste'able with your suggestion (Will error if ran).
My approach also makes the value easily select'able remove'able since it's considered one word _(Alt+Backspace removes everything and double clicking selects everything)_

It's not a super big deal, so I can definitely change if you prefer the other way. But I would like your thoughts on it

Reviewer: You seem to have put way more thought into this than me. :smile: Can we then have it like:

```suggestion
export CK8S_PGP_FP="FOOBAR1234567" # retrieve with gpg --list-secret-keys
```

export CK8S_CLOUD_PROVIDER=baremetal # We don't have a OVH flavor, but baremetal is fine
export CK8S_FLAVOR=dev # Change to "prod" if it's a production cluster you're setting up

S3_ACCESS_KEY="foo" # Access key for S3, see https://docs.ovh.com/gb/en/public-cloud/getting_started_with_the_swift_S3_API/#create-ec2-credentials
S3_SECRET_KEY="bar" # Secret key for S3
```

Download the kubeconfig and set them up for Compliant Kubernetes by following these steps:

> 1. Create the path where they're going to be stored

Reviewer: Can you change these to heading 3? I find it clumpy to indent code blocks and started moving way from that in Exoscale.

    ```ShellSession
    mkdir -p "${CK8S_CONFIG_PATH}/.state"
    ```

1. Download the kubeconfig from OVH for the service cluster and run:

    ```ShellSession
    mv ~/Downloads/kubeconfig.yml "${CK8S_CONFIG_PATH}/.state/kube_config_sc.yaml"

>     sops --encrypt --in-place --pgp "${CK8S_PGP_FP}" "${CK8S_CONFIG_PATH}/.state/kube_config_sc.yaml"

Reviewer: You probably thought this through, but just to make sure.

This goes against [ADR-0006 Use standard Kubeconfig mechanisms](https://compliantkubernetes.io/adr/0006-use-standard-kubeconfig-mechanisms/). Do I get correctly that you consider adding a way to use KUBECONFIG and kubecontext in a standard way a task too big to piggyback it onto this task?

Author: Since apps wont read the kubeconfig as long as it's not encrypted and in this location I don't see any other way around this issue.

> 1. Download the kubeconfig from OVH for the SC cluster and run:
> 
>     ```ShellSession
>     mv ~/Downloads/kubeconfig.yml "${CK8S_CONFIG_PATH}/.state/kube_config_sc.yaml"
>     sops --encrypt --in-place --pgp "${CK8S_PGP_FP}" "${CK8S_CONFIG_PATH}/.state/kube_config_${SERVICE_CLUSTER}.yaml"
>     ```
> 
> 1. Download the kubeconfig from OVH for the wc cluster and run:
> 
>     ```ShellSession
>     mv ~/Downloads/kubeconfig.yml "${CK8S_CONFIG_PATH}/.state/kube_config_wc.yaml"
>     sops --encrypt --in-place --pgp "${CK8S_PGP_FP}" "${CK8S_CONFIG_PATH}/.state/kube_config_wc.yaml"
>     ```

Reviewer 2: ```suggestion
1. Download the kubeconfig from OVH for the service cluster and run:

    ```ShellSession
    mv ~/Downloads/kubeconfig.yml "${CK8S_CONFIG_PATH}/.state/kube_config_sc.yaml"
    sops --encrypt --in-place --pgp "${CK8S_PGP_FP}" "${CK8S_CONFIG_PATH}/.state/kube_config_sc.yaml"
    ```

1. Download the kubeconfig from OVH for the workload cluster and run:

    ```ShellSession
    mv ~/Downloads/kubeconfig.yml "${CK8S_CONFIG_PATH}/.state/kube_config_wc.yaml"
    sops --encrypt --in-place --pgp "${CK8S_PGP_FP}" "${CK8S_CONFIG_PATH}/.state/kube_config_wc.yaml"
    ```
```

Set up DNS records for the following subdomains

> ```ShellSession

Reviewer: I'm not sure what OVH uses, but can this be something that is copy-paste-able in OVH's DNS, as we did for AWS's Route53? Consider buying a new domain if it makes your life easier.

Author: I didn't look into it. And I think you need to do it via their GUI since although they have a CLI it's depricated and not maintained anymore.
But I'll see if maybe they already have their own guide for this or otherwise I can add take some screenshots and add it here

Reviewer: They seems to have a GUI option "Change in text format": https://docs.ovh.com/us/en/domains/web_hosting_how_to_edit_my_dns_zone/. Can you explore that a bit? It feels more precise to copy-paste a DNS zone file, instead of suggesting many manual click-a-de-click edits.

Author: Absolutely, I'll look into that

Create required buckets that Compliant Kubernetes will use.

```ShellSession
cat <<EOF | sops --encrypt --pgp "${CK8S_PGP_FP}" --output-type ini --input-type ini /dev/stdin > s3cmd.ini
[default]
use_https   = True
host_base   = s3.${REGION}.cloud.ovh.net
host_bucket = s3.${REGION}.cloud.ovh.net
access_key  = ${S3_ACCESS_KEY}
secret_key  = ${S3_SECRET_KEY}
EOF

sops exec-file --no-fifo s3cfg.ini "s3cmd --config {} mb s3://${CK8S_ENVIRONMENT_NAME}-harbor"
sops exec-file --no-fifo s3cfg.ini "s3cmd --config {} mb s3://${CK8S_ENVIRONMENT_NAME}-velero"
sops exec-file --no-fifo s3cfg.ini "s3cmd --config {} mb s3://${CK8S_ENVIRONMENT_NAME}-es-backup"
sops exec-file --no-fifo s3cfg.ini "s3cmd --config {} mb s3://${CK8S_ENVIRONMENT_NAME}-influxdb"
sops exec-file --no-fifo s3cfg.ini "s3cmd --config {} mb s3://${CK8S_ENVIRONMENT_NAME}-sc-logs"
```

Download Compliant Kubernetes and checkout the latest version.

```ShellSession
git clone git@github.com:elastisys/compliantkubernetes-apps.git
cd compliantkubernetes-apps
git checkout v0.11.0
```

Initialize the config.

```ShellSession
bin/ck8s init
```

Update the service cluster configuration for OVH

> ```ShellSession

Reviewer: Nice!

yq write --inplace wc-config.yaml global.issuer "letsencrypt-prod"

yq write --inplace wc-config.yaml storageClasses.default "csi-cinder-high-speed"
yq write --inplace wc-config.yaml storageClasses.local.enabled "false"

yq write --inplace wc-config.yaml objectStorage.s3.region "${REGION}"
yq write --inplace wc-config.yaml objectStorage.s3.regionAddress "s3.${REGION}.cloud.ovh.net"
yq write --inplace wc-config.yaml objectStorage.s3.regionEndpoint "https://s3.${REGION}.cloud.ovh.net/"

yq write --inplace wc-config.yaml ingressNginx.controller.useHostPort "false"
yq write --inplace wc-config.yaml ingressNginx.controller.service.enabled "true"
yq write --inplace wc-config.yaml ingressNginx.controller.service.type "LoadBalancer"
yq write --inplace wc-config.yaml ingressNginx.controller.service.annotations ""

yq write --inplace wc-config.yaml issuers.letsencrypt.prod.email "${ISSUER_MAIL}"
yq write --inplace wc-config.yaml issuers.letsencrypt.staging.email "${ISSUER_MAIL}"

yq write --inplace wc-config.yaml metricsServer.enabled "false"
```

Set s3 credentials

```ShellSession
sops --set '["objectStorage"]["s3"]["accessKey"] "${S3_ACCESS_KEY}"' secrets.yaml

> sops --set '["objectStorage"]["s3"]["secretKey"] "${S3_SECRET_KEY"' secrets.yaml

Reviewer 2: ```suggestion
sops --set '["objectStorage"]["s3"]["secretKey"] "${S3_SECRET_KEY}"' secrets.yaml
```

> sops --set '["objectStorage"]["s3"]["secretKey"] "${S3_SECRET_KEY"' secrets.yaml

Reviewer 2: Will `s3_SECRET_KEY` be interpreted here? I'm unsure since it's within ' '

Author: You are correct, it will not.

```

### Deploy Compliant Kubernetes

Now you're ready to deploy Compliant Kubernetes.
When the apply command is done, fetch the external IP assigned to the loadbalancer service and update the DNS record to match this.

```ShellSession
bin/ck8s apply sc

bin/ck8s ops kubectl sc get svc -n ingress-nginx ingress-nginx-controller
```

Update the following records to match the loadbalancer IP in the service cluster

> ```ShellSession

Reviewer: Same comment as above. If this could produce something copy-paste-able to OVH's DNS, it would be great!

          aws_key_id "#{ENV['AWS_ACCESS_KEY_ID']}"
          aws_sec_key "#{ENV['AWS_ACCESS_SECRET_KEY']}"
          s3_endpoint https://s3.waw.cloud.ovh.net/ # <--- Add this line
          s3_region waw
        .
        .
    $ bin/ck8s ops kubectl sc delete pod -n fluentd fluentd-0
    ```

1. Update velero to use 1.3.0 instead of 1.2.0, _fixed with [this issue](https://github.com/elastisys/compliantkubernetes-apps/issues/289)_

    For velero to work with OVH S3, velero needs to run with version `1.3.0`.

    ```ShellSession
    bin/ck8s ops kubectl sc patch deployment -n velero velero -p '{"spec":{"template":{"spec":{"containers":[{"name":"velero","image":"velero/velero:v1.3.0"}]}}}}'
    bin/ck8s ops kubectl sc patch daemonset -n velero restic -p '{"spec":{"template":{"spec":{"containers":[{"name":"velero","image":"velero/velero:v1.3.0"}]}}}}'

    bin/ck8s ops kubectl wc patch deployment -n velero velero -p '{"spec":{"template":{"spec":{"containers":[{"name":"velero","image":"velero/velero:v1.3.0"}]}}}}'
    bin/ck8s ops kubectl wc patch daemonset -n velero restic -p '{"spec":{"template":{"spec":{"containers":[{"name":"velero","image":"velero/velero:v1.3.0"}]}}}}'

>     ```

Reviewer: This is such a cool approach between "we need to wait until everything is fixed" and "we have no clue". Thanks for championing it!
